pkg/resourcemanager/cosmosdb/sqldatabase/cosmosdbsql.go

```diff
@@ -0,0 +1,275 @@
+	}
+
+	notFoundErrorCodes := []string{
+		errhelp.ParentNotFoundErrorCode,
+		errhelp.ResourceGroupNotFoundErrorCode,
+		errhelp.NotFoundErrorCode,
+		errhelp.ResourceNotFound,
+	}
+
+	// Get the existing resource
+	sqlDatabase, err := cosmosDBSQLDatabaseClient.GetSQLDatabase(
+		ctx,
+		instance.Spec.ResourceGroup,
+		instance.Spec.CosmosDBAccount,
+		instance.ObjectMeta.Name)
+	if err != nil {
+		azerr := errhelp.NewAzureError(err)
+		if azerr.Type != errhelp.ResourceNotFound && azerr.Type != errhelp.NotFoundErrorCode {
+			instance.Status.Message = fmt.Sprintf("failed GET-ing CosmosDBSQLDatabase: %s", err.Error())
+			return errhelp.IsErrorFatal(err, notFoundErrorCodes, nil)
+		}
+	}
+
+	// TODO: There's a bug in Cosmos DB SQL Database where the throughput settings configured on
+	// TODO: the DB don't actually get shown on the DB, instead they are shown on the throughput child resource.
+	// TODO: We monitor that resource too so we can diff against it
+	sqlDatabaseThroughputSettings, err := cosmosDBSQLDatabaseClient.GetSQLDatabaseThroughput(
+		ctx,
+		instance.Spec.ResourceGroup,
+		instance.Spec.CosmosDBAccount,
+		instance.ObjectMeta.Name)
+	if err != nil {
+		azerr := errhelp.NewAzureError(err)
+		if azerr.Type != errhelp.ResourceNotFound && azerr.Type != errhelp.NotFoundErrorCode {
+			instance.Status.Message = fmt.Sprintf("failed GET-ing CosmosDBSQLDatabaseThroughput: %s", err.Error())
+			return errhelp.IsErrorFatal(err, notFoundErrorCodes, nil)
+		}
+	}
+
+	// Transform our spec to Azure shape
+	transformed := m.transformToCosmosDBSQLDatabase(instance)
+
+	// Check if we're in the goal state already
+	resourceMatchesAzure := DoesResourceMatchAzure(transformed, sqlDatabase, sqlDatabaseThroughputSettings)
+	if resourceMatchesAzure {
+		instance.Status.SetProvisioned(resourcemanager.SuccessMsg)
+		instance.Status.ResourceId = *sqlDatabase.ID
+		return true, nil
+	}
+
+	// Drive to goal state
+	instance.Status.SetProvisioning("")
+	future, err := cosmosDBSQLDatabaseClient.CreateUpdateSQLDatabase(
+		ctx,
+		instance.Spec.ResourceGroup,
+		instance.Spec.CosmosDBAccount,
+		instance.ObjectMeta.Name,
+		transformed)
+	if err != nil {
+		instance.Status.Message = err.Error()
+		allowedErrors := []string{
+			errhelp.AsyncOpIncompleteError,
+			errhelp.AlreadyExists,
+			errhelp.FailoverGroupBusy,
+		}
+		unrecoverableErrors := []string{
+			errhelp.InvalidFailoverGroupRegion,
+		}
+		return errhelp.IsErrorFatal(err, append(allowedErrors, notFoundErrorCodes...), unrecoverableErrors)
+	}
+	instance.Status.SetProvisioning("Resource request successfully submitted to Azure")
+	instance.Status.PollingURL = future.PollingURL()
+
+	// Need to poll the polling URL, so not done yet!
+	return false, nil
+}
+
+func (m *AzureCosmosDBSQLDatabaseManager) Delete(ctx context.Context, obj runtime.Object, opts ...resourcemanager.ConfigOption) (bool, error) {
+	options := &resourcemanager.Options{}
+	for _, opt := range opts {
+		opt(options)
+	}
```

Comment:
`options` isn't used in this method - is it just for side-effects in the ConfigOption callbacks?

Reply:
No it was just boilerplate from other controllers. I've removed the `options` construction for now and left `opts` unused.